File: functions/groups/create_chat_message.py
from firebase_admin import firestore
from firebase_admin.firestore import SERVER_TIMESTAMP
from flask import Request, abort
from models.constants import Collections, GroupFields, ChatFields
from models.data_models import ChatMessage
from utils.logging_utils import get_logger


def create_group_chat_message(request: Request, group_id: str) -> ChatMessage:
    """
    Creates a new chat message in a specific group.

    This function adds a new message to the group's chats subcollection.

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
                - validated_params: The validated request body containing:
                    - text: The message text
                    - attachments: Optional list of attachment URLs
        group_id: The ID of the group to add the chat message to

    Returns:
        A ChatMessage object representing the newly created message

    Raises:
        404: Group not found
        403: User is not a member of the group
        400: Invalid request data
        500: Internal server error
    """
    logger = get_logger(__name__)
    logger.info(f"Creating new chat message in group: {group_id}")

    # Get the authenticated user ID from the request
    current_user_id = request.user_id

    # Get the validated request data
    validated_data = request.validated_data
    text = validated_data.text
    attachments = validated_data.attachments or []

    # Initialize Firestore client
    db = firestore.client()

    # First, check if the group exists and if the user is a member
    group_ref = db.collection(Collections.GROUPS).document(group_id)
    group_doc = group_ref.get()

    if not group_doc.exists:
        logger.warning(f"Group {group_id} not found")
        abort(404, description="Group not found")

    group_data = group_doc.to_dict()
    members = group_data.get(GroupFields.MEMBERS, [])

    # Check if the current user is a member of the group
    if current_user_id not in members:
        logger.warning(f"User {current_user_id} is not a member of group {group_id}")
        abort(403, description="You must be a member of the group to post messages")

    # Create the chat message
    chats_ref = group_ref.collection(Collections.CHATS)

    # Prepare the message data
    message_data = {
        ChatFields.SENDER_ID: current_user_id,
        ChatFields.TEXT: text,
        ChatFields.CREATED_AT: SERVER_TIMESTAMP,
    }

    # Add attachments if provided
    if attachments:
        message_data[ChatFields.ATTACHMENTS] = attachments

    # Add the message to Firestore
    new_message_ref = chats_ref.document()  # Auto-generate ID
    new_message_ref.set(message_data)

    # Get the created message
    new_message_doc = new_message_ref.get()
    new_message_data = new_message_doc.to_dict()

    # Convert server timestamp to string for the response
    created_at = ""
    if new_message_data.get(ChatFields.CREATED_AT):
        created_at = new_message_data.get(ChatFields.CREATED_AT).isoformat()

    logger.info(f"Created new chat message with ID: {new_message_ref.id}")

    # Return the created message
    return ChatMessage(
        message_id=new_message_ref.id,
        sender_id=current_user_id,
        text=text,
        created_at=created_at,
        attachments=attachments,
    )

File: functions/src/groups/create-chat-message.ts
import { Request, Response } from 'express';
import { getFirestore, Timestamp } from 'firebase-admin/firestore';
import { Collections } from '../models/constants.js';
import { ChatMessage, CreateChatMessagePayload } from '../models/data-models.js';
import { chatConverter, ChatDoc } from '../models/firestore/chat-doc.js';
import { groupConverter } from '../models/firestore/group-doc.js';
import { ForbiddenError, NotFoundError } from '../utils/errors.js';
import { getLogger } from '../utils/logging-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Creates a new chat message in a specific group.
 *
 * This function adds a new message to the group's chats subcollection.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: The validated request body containing:
 *                - text: The message text (required)
 *                - attachments: Optional list of attachment URLs
 * @param res - The Express response object
 * @param groupId - The ID of the group to add the chat message to
 *
 * @returns A ChatMessage object representing the newly created message
 *
 * @throws 404: Group not found
 * @throws 403: User is not a member of the group
 * @throws 400: Invalid request data (missing text)
 * @throws 500: Internal server error
 */
export const createGroupChatMessage = async (req: Request, res: Response, groupId: string): Promise<void> => {
  logger.info(`Creating new chat message in group: ${groupId}`);

  // Get the authenticated user ID from the request
  const currentUserId = req.userId;

  // Get the validated request data
  const validatedParams = req.validated_params as CreateChatMessagePayload;
  const text = validatedParams.text;
  const attachments = validatedParams.attachments ?? [];

  // Initialize Firestore client
  const db = getFirestore();

  // First, check if the group exists and if the user is a member
  const groupRef = db.collection(Collections.GROUPS).withConverter(groupConverter).doc(groupId);
  const groupDoc = await groupRef.get();

  if (!groupDoc.exists) {
    logger.warn(`Group ${groupId} not found`);
    throw new NotFoundError('Group not found');
  }

  const groupData = groupDoc.data();
  const members = groupData?.members || [];

  // Check if the current user is a member of the group
  if (!members.includes(currentUserId)) {
    logger.warn(`User ${currentUserId} is not a member of group ${groupId}`);
    throw new ForbiddenError('You must be a member of the group to post messages');
  }

  // Create the chat message
  const chatsRef = db
    .collection(Collections.GROUPS)
    .doc(groupId)
    .collection(Collections.CHATS)
    .withConverter(chatConverter);

  const currentTime = Timestamp.now();

  // Prepare the message data
  const messageData: ChatDoc = {
    sender_id: currentUserId,
    text: text,
    created_at: currentTime,
    attachments: attachments.map((url) => ({ type: 'image', url })),
  };

  // Add the message to Firestore
  const newMessageRef = chatsRef.doc(); // Auto-generate ID
  await newMessageRef.set(messageData);

  logger.info(`Created new chat message with ID: ${newMessageRef.id}`);

  // Return the created message
  const response: ChatMessage = {
    message_id: newMessageRef.id,
    sender_id: currentUserId,
    text,
    created_at: formatTimestamp(currentTime),
    attachments,
  };

  res.json(response);
};
